master: Add python_requires to help pip

hugovk opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Helps towards #17943 "Removing Python 2 code".

Like https://github.com/sympy/sympy/pull/17966 but for `master`.

#### Brief description of what is fixed or changed

Add `python_requires` to help pip install the right version of this package for the user's running Python.

Also update the Trove classifiers.

Also update the version check in `setup.py`, Python 3.4 was already dropped in https://github.com/sympy/sympy/pull/16492. This is not strictly necessary with `python_requires`, unless the user has a really old version of pip and could probably be removed entirely.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

[I'm not sure if this requires user-facing release notes, so didn't add any, but please let me know if otherwise.]


<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

hugovk committed:
Update version check, Trove classifiers, and add python_requires to help pip

oscarbenjamin commented:
This looks good but this won't be mergeable until we drop Python 2.7 from Travis. I think we've agreed to hold off on that for a little while just until we get feedback from the SymPy 1.5 release.

kenluck2001 commented on sympy/__init__.py:
```diff
@@ -35,10 +35,8 @@ def enable_warnings():
-    raise ImportError("Python version 2.7 or 3.4 or above "
-                      "is required for SymPy.")
+if sys.version_info < (3, 5):
+    raise ImportError("Python 3.5 or above is required for SymPy.")
```
Are you making a backward-incompatible change?

hugovk replied:
Yes, the idea is to drop support for Python 2.7: https://github.com/sympy/sympy/issues/17943.

asmeurer replied:
1.5.1 will be the last version to support Python 2. See https://github.com/sympy/sympy/wiki/Python-version-support-policy

hugovk commented:
Rebased to resolve conflicts.

oscarbenjamin commented:
I guess this needs a change to the `TEST_PY2_IMPORT` case in `test_travis.sh`. Maybe we can test importing sympy before trying to install it.

oscarbenjamin commented:
I have a fix for this in #18244 (changes to `.travis.yml` and `bin/test_py2_import`). That fix is needed or we also can't have any Py3 syntax in the main codebase...

oscarbenjamin commented:
It might make sense to extract those changes and push them here

hugovk committed:
Fix TEST_PY2_IMPORT

hugovk committed:
sudo no longer needed https://blog.travis-ci.com/2018-11-19-required-linux-infrastructure-migration

hugovk commented:
Thanks, updated, and `TEST_PY2_IMPORT` now passes!

Also removed the redundant `sudo:` lines from `.travis.yml`.

* https://blog.travis-ci.com/2018-11-19-required-linux-infrastructure-migration

czgdp1807 commented:
@hugovk Please resolve this conflicts. Probably after that it should be good to go as can be concluded from the the above comments.

hugovk committed:
Merge branch 'master' into master-add-python_requires

hugovk commented:
@czgdp1807 Conflict resolved!

asmeurer commented:
What is the purpose of the `sudo` changes in .travis.yml?

Other than that, the rest looks good to me.

hugovk commented:
@asmeurer `sudo` no longer has any effect:

* [blog.travis-ci.com/2018-11-19-required-linux-infrastructure-migration](https://blog.travis-ci.com/2018-11-19-required-linux-infrastructure-migration)

asmeurer commented:
I guess it works without it.
